irispy/spectrograph.py

```diff
@@ -61,132 +50,46 @@ def __init__(self, filenames, spectral_windows="All"):
-                             "time step size mean": hdulist[0].header["STEPT_AV"],
-                             "time step size sigma": hdulist[0].header["STEPT_DV"],
-                             "spectral windows in OBS": windows_in_obs,
-                             "spectral windows in object": spectral_windows,
-                             "detector gain": iris_tools.DETECTOR_GAIN,
-                             "detector yield": iris_tools.DETECTOR_YIELD,
-                             "readout noise": iris_tools.READOUT_NOISE}
-                # Translate some metadata to be more helpful.
-                if hdulist[0].header["IAECEVFL"] == "YES":
-                    self.meta["IAECEVFL"] = True
-                if hdulist[0].header["IAECFLAG"] == "YES":
-                    self.meta["IAECFLAG"] = True
-                if hdulist[0].header["IAECFLFL"] == "YES":
-                    self.meta["IAECFLFL"] = True
-                if self.meta["data level"] == 2.:
-                    if self.meta["camera"] == 1:
-                        self.meta["camera"] = "spectra"
-                    elif self.meta["camera"] == 2:
-                        self.meta["camera"] = "SJI"
-                # Define empty dictionary with keys corresponding to
-                # spectral windows.  The value of each key will be a
-                # list of xarray data arrays, one for each raster scan.
-                data_dict = dict([(window_name, None) for window_name in self.spectral_windows["name"]])
-                # Record header info of auxiliary data.  Should be
-                # consistent between files of same OBS.
+                # creating a empty list for every spectral window and each spectral window
+                # is a key for the dictionary.
+                data_dict = dict([(window_name, list())
+                                  for window_name in self.spectral_windows["name"]])
                 auxiliary_header = hdulist[-2].header
-            # Extract the data and meta/auxiliary data.
-            # Create WCS object from FITS header and add WCS object
-            # wcs dictionary.
-            wcs_celestial = wcs.WCS(hdulist[1].header).celestial
+            # the unchanged header of the hdulist indexed 0.
+            self.meta = hdulist[0].header
+            for i, window_name in enumerate(self.spectral_windows["name"]):
+                wcs_ = WCS(hdulist[window_fits_indices[i]].header)
+                data_ = hdulist[window_fits_indices[i]].data
+                data_nan_masked = copy.deepcopy(hdulist[window_fits_indices[i]].data)
+                data_nan_masked[hdulist[window_fits_indices[i]].data == -200.] = np.nan
+                # appending Cube instance to the corresponding window key in dictionary's list.
+                data_dict[window_name].append(
+                    Cube(data_, wcs_, meta=dict(self.meta), mask=data_nan_masked))
```

Comment:
So the ```data_nan_masked``` isn't the mask itself.  It's the data with all pixels with no data (i.e. pixels = -200) set to a value of ```np.nan```.  The mask should be ```data_mask = hdulist[window_fits_indices[i]].data == -200.```.  Therefore, this line should be
```data_dict[window_name].append(Cube(data_nan_masked, wcs_, meta=dict(self.meta), mask=hdulist[window_fits_indices[i]].data == -200.))```
this also means line ```data_ = hdulist[window_fits_indices[i]].data``` can be removed.